design-documents/graph-ql/framework/async-mutations.md

```diff
@@ -0,0 +1,281 @@
+Command that has been invoked and did not complete yet will be marked as *in progress*.
+
+```graphql schema
+type Query {
+    queue(operationIdentifiers: [String!]!) : [OperationStateInterface]
+}
+
+type OperationCompleted implements OperationStateInterface {
+     identifier: String!
+}
+
+type OperationInProgress implements OperationStateInterface {
+     identifier: String!
+}
+```
+
+#### Request
+```graphql
+query {
+    queue($operationIdentifier: [String!]!) {
+        identifier
+        __typename__
+    }
+}
+```
+#### Response
+```json
+{  
+   "queue":[  
+      {  
+         "identifier":"21e0e65f29bcaf513d2685815916ab7e5ac9d473",
+         "__typename__":"OperationAccepted"
+      },
+      {  
+         "identifier":"95f7cf4623b9a53171e77caffd05b0e54cdc11b3",
+         "__typename__":"OperationInProgress"
+      },
+      {  
+         "identifier":"e211fb19a5ce81969aa5ec3ab216d2401c8766ed",
+         "__typename__":"OperationCompleted"
+      }
+   ]
+}
+```
+
+### Handling failures
+
+In case of failure operation status will be changed in a way to provide client information about the cause of the issue.
+
+```graphql schema
+interface FailureReportInterface {
+    message: String!
+}
+
+type LogicalErrorReport implements FailureReportInterface {
+    message: String!
+}
+
+type OperationFailed implements OperationStateInterface {
+     identifier: String!
+     failures: [FailureReportInterface!]!
+}
+
+```
+#### Request
+```graphql 
+query {
+    queue($operataionIdentifier: [String!]) {
+        identifier
+        __typename__
+        ... on OperationFailed {
+            failures {
+                message
+                __typename__
+            }
+        }
+    }
+}
+```
+#### Response
+```json
+{
+   "queue":[
+      {
+         "identifier":"25d7d02fa2b73238b098b0254d9ec98a5cd23696",
+         "__typename__":"OperationFailed",
+         "failures":[
+            {
+               "__typename__":"LogicalErrorReport",
+               "message":"No such entity with email example@example.com"
+            }
+         ]
+      }
+   ]
+}
+```
+
+### Retrieving results
+
+Not all of commands require execution results.
+For some operations this is enough to return just an execution status.
+In case the application requires to work with the mutation execution results exists two ways for this. Generic and custom.
+
+#### Generic
+
+Generic way provides a possibility to return with success list of types that were changes during mutation execution.
+Basically, this is a list of updated entities which should be reloaded for application work. The application can use this result as tags for invalidation local cached entities etc.
+Filling of this list is a developer responsibility and does not guarantee by the framework.
+To return information about such entities operation state must implement 
+
+```graphql schema
+type ChangedEntity {
+    type: String!
+    identifiers: [String!]
```

Comment:
To me personally current option seems more complicated than flat structure of changed entities.
```suggestion
    identifier: String!
```